### ISSUE TITLE
[Feature Request] Vuetify 3 + VitePress

### ISSUE THREAD

**Issue by anisabboud** [vuetifyjs/vuetify#13859]
### Problem to solve
Can Vuetify 3 be used with VitePress?

What steps to follow? Thanks!

**Comment by KaelWD** [comment]
Probably. 

The issue list of this repo is **exclusively** for bug reports and feature requests. For general questions, please join [the Discord chat room](https://community.vuetifyjs.com). You can also check [reddit](https://www.reddit.com/r/vuetifyjs/) or [stackoverflow](https://stackoverflow.com/questions/tagged/vuetify.js).

**Comment by anisabboud** [comment]
This is a feature request. The [Vuetify 3 Installation page](https://next.vuetifyjs.com/en/getting-started/installation/) explains how to use it with Vite, but doesn't mention using it with VitePress.

Since both VitePress and Vuetify 3 are still in alpha stages, I don't think the stackoverflow community can answer this issue.
Also, when someone asked this on [stackoverflow](https://stackoverflow.com/questions/67415473/how-to-use-vuetify-3-in-vitepress), the question was [marked off-topic, then closed and deleted](https://webcache.googleusercontent.com/search?q=cache:EjqL3vwoKNcJ:https://stackoverflow.com/questions/67415473/how-to-use-vuetify-3-in-vitepress+).

VitePress and Vuetify 3 are two major Vue projects that are expected to gain more popularity once they're stable.
I believe many people would want to use them in conjunction.

**Comment by dulimarta** [comment]
I share @anisabboud request. For years, I have been using VuePress with custom components built using Vuetify 2.x and I'd like to upgrade my setup to VitePress and Vuetify 3.x. I tried the following in my `.vitepress/theme/index.ts`

```
import DefaultTheme from "vitepress/theme";
import "vuetify/styles";
import { createVuetify } from "vuetify";

const vuetify = createVuetify();
export default {
  ...DefaultTheme,
  enhanceApp({ app, siteData }) {
    app.use(vuetify);
  },
};
```

But got error messages related to unregisterd Vuetify component (`v-card`, `v-btn`, etc)

**Comment by KaelWD** [comment]
You need to import components too
```js
import * as components from 'vuetify/components'
createVuetify({ components })
``
Or use https://github.com/vuetifyjs/vuetify-loader/tree/next/packages/vite-plugin#automatic-imports

**Comment by dulimarta** [comment]
@KaelWD thank you so much for your prompt response. I would never expect to get a response within 18 minutes of my post.

Your answer solves my problem. I'm hoping the other people having the same issue would see your answer above. After adding  hose two linesm,  error messages on unregistered Vuetify components are gone. Vitepress now recognizes them, what I have to do next is to migrate from Vuetify 2.x to Vuetify 3.x syntax.

**Comment by dulimarta** [comment]
I have been having issue using vuetify-3.0.0-beta-10 and vitepress-1.0.0-alpha.13. Everything seems to be working when `vitepress dev` but got the following error when running `vitepress build`:

```
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".css" for C:\xxxxxxxxxx\node_modules\vuetify\lib\components\VCode\VCode.css
    at new NodeError (node:internal/errors:371:5)
    at Object.file: (node:internal/modules/esm/get_format:72:15)
    at defaultGetFormat (node:internal/modules/esm/get_format:85:38)
    at defaultLoad (node:internal/modules/esm/load:13:42)
    at ESMLoader.load (node:internal/modules/esm/loader:303:26)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:230:58)
    at new ModuleJob (node:internal/modules/esm/module_job:63:26)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:244:11)
    at async ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:78:21)
    at async Promise.all (index 0) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}
```

**Comment by ctrueblood-epri** [comment]
@dulimarta, I encountered the same error during a SSR build: *Unknown file extension ".css"*. Please see #15700 for a possible solution.

**Comment by andjimrio** [comment]
I do that in my `.vitepress/theme/index.ts` according to [Vuetify docs](https://vuetifyjs.com/en/getting-started/installation/#existing-projects)

```js
import { h } from 'vue'
import type { Theme } from 'vitepress'
import DefaultTheme from 'vitepress/theme'
import './style.css'

import 'vuetify/styles'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'

const vuetify = createVuetify({
  components,
  directives,
})

export default {
  extends: DefaultTheme,
  Layout: () => {
    return h(DefaultTheme.Layout, null, {})
  },
  enhanceApp({ app, router, siteData }) {
    app.use(vuetify)
  }
} satisfies Theme

```

**Comment by CrazyOrr** [comment]
> I do that in my `.vitepress/theme/index.ts` according to [Vuetify docs](https://vuetifyjs.com/en/getting-started/installation/#existing-projects)
> 
> ```js
> import { h } from 'vue'
> import type { Theme } from 'vitepress'
> import DefaultTheme from 'vitepress/theme'
> import './style.css'
> 
> import 'vuetify/styles'
> import { createVuetify } from 'vuetify'
> import * as components from 'vuetify/components'
> import * as directives from 'vuetify/directives'
> 
> const vuetify = createVuetify({
>   components,
>   directives,
> })
> 
> export default {
>   extends: DefaultTheme,
>   Layout: () => {
>     return h(DefaultTheme.Layout, null, {})
>   },
>   enhanceApp({ app, router, siteData }) {
>     app.use(vuetify)
>   }
> } satisfies Theme
> ```
> 
> But... I have a problem. The styles are not correctly import. Everything works expect the css classes because they are override by the vitepress' styles. @KaelWD , any idea?

Use [`vp-raw`](https://github.com/vuejs/vitepress/discussions/3578#discussioncomment-8495967) class to remove default styles.

**Comment by andjimrio** [comment]
> Use [`vp-raw`](https://github.com/vuejs/vitepress/discussions/3578#discussioncomment-8495967) class to remove default styles.

@CrazyOrr, Can you explain better? Where should I insert `vp-raw`?

**Comment by CrazyOrr** [comment]
> > Use [`vp-raw`](https://github.com/vuejs/vitepress/discussions/3578#discussioncomment-8495967) class to remove default styles.
> 
> @CrazyOrr, Can you explain better? Where should I insert `vp-raw`?

Wrap your component with [`raw` container](https://vitepress.dev/guide/markdown#raw) 
```md
::: raw
<MyComponent />
:::
```
or directly add `vp-raw` class to it (or any of its parents)
```vue
<MyComponent  class="vp-raw"/>
```
, then apply `postcss` (install it if you haven't) [transformations](https://github.com/vuejs/vitepress/blob/8aa6ccbe32655f76c390d15568f69f83d079385d/src/node/postcss/isolateStyles.ts) to those [css files](https://github.com/vuejs/vitepress/tree/main/src/client/theme-default/styles) that contain styles you want to remove (typically [`base.css`](https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/styles/base.css), [`vp-doc.css`](https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/styles/components/vp-doc.css)).
```js
// postcss configuration file: postcss.config.mjs
import { postcssIsolateStyles } from 'vitepress';

export default {
  plugins: [
    postcssIsolateStyles({
      includeFiles: [/base\.css/, /vp-doc\.css/],
    }),
  ],
};
```
Basically, `.xxx` became `.xxx:not(:where(.vp-raw, .vp-raw *))` so that it no longer take effect under `vp-raw` class.

**Comment by betaglot** [comment]
@CrazyOrr What directory does the postcss configuration file go in?

**Comment by CrazyOrr** [comment]
> [@CrazyOrr](https://github.com/CrazyOrr) What directory does the postcss configuration file go in?

In [`./` or `./docs`](https://vitepress.dev/guide/getting-started#file-structure).

**Comment by CrazyOrr** [comment]
No problem.

VitePress has a hook [`useData`](https://vitepress.dev/reference/runtime-api#usedata) which returns page-specific data, which contains a field `isDark`. You can [watch](https://vuejs.org/guide/essentials/watchers.html) this field to [change the theme](https://vuetifyjs.com/en/features/theme/#changing-theme) of the Vuetify components accordingly.

Example:
```js
import { useData } from 'vitepress';
import { useTheme } from 'vuetify';
import { watch } from 'vue';

const { isDark } = useData();
const theme = useTheme();
watch(
  isDark,
  (isDark) => {
    theme.global.name.value = isDark ? 'dark' : 'light';
  },
  { immediate: true },
);
```
